libtransmission/tr-udp.cc

```diff
@@ -92,51 +95,44 @@ static void set_socket_buffers(tr_socket_t fd, bool large)
     }
 }
 
-void tr_udpSetSocketBuffers(tr_session const* session)
+void tr_session::tr_udp_core::set_socket_buffers()
 {
-    bool const utp = session->allowsUTP();
+    bool const utp = session_->allowsUTP();
 
-    if (session->udp_socket != TR_BAD_SOCKET)
+    if (udp_socket_ != TR_BAD_SOCKET)
     {
-        set_socket_buffers(session->udp_socket, utp);
+        ::set_socket_buffers(udp_socket_, utp);
     }
-
-    if (session->udp6_socket != TR_BAD_SOCKET)
+    if (udp6_socket_ != TR_BAD_SOCKET)
     {
-        set_socket_buffers(session->udp6_socket, utp);
+        ::set_socket_buffers(udp6_socket_, utp);
     }
 }
 
-void tr_udpSetSocketTOS(tr_session* session)
-{
-    session->setSocketTOS(session->udp_socket, TR_AF_INET);
-    session->setSocketTOS(session->udp6_socket, TR_AF_INET6);
-}
-
 /* BEP-32 has a rather nice explanation of why we need to bind to one
    IPv6 address, if I may say so myself. */
 // TODO: remove goto, it prevents reducing scope of local variables
-static void rebind_ipv6(tr_session* ss, bool force)
+void tr_session::tr_udp_core::rebind_ipv6(bool force)
 {
     struct sockaddr_in6 sin6;
-    unsigned char const* ipv6 = tr_globalIPv6(ss);
+    unsigned char const* ipv6 = tr_globalIPv6(session_);
     int rc = -1;
     int one = 1;
 
     /* We currently have no way to enable or disable IPv6 after initialisation.
        No way to fix that without some surgery to the DHT code itself. */
-    if (ipv6 == nullptr || (!force && ss->udp6_socket == TR_BAD_SOCKET))
+    if (ipv6 == nullptr || (!force && udp6_socket_ == TR_BAD_SOCKET))
     {
-        if (ss->udp6_bound != nullptr)
+        if (udp6_bound_ != nullptr)
         {
-            free(ss->udp6_bound);
-            ss->udp6_bound = nullptr;
+            free(udp6_bound_);
```

Comment:
Optional suggestion if you're interested: except for the third-party `jsonsl` code, this is the _only remaining code_ that uses `malloc()` and `free()` in libtransmission and it took a _lot of work_ to get here. Feel free either in this PR or a followup to change this type to something else, e.g. `std::optional<decltype(sockaddr_in6::sin6_addr)>`  and add `cppcoreguidelines-no-malloc` to `libtransmission/.clang-tidy` :partying_face:  

If you want to go a step further with that and change the return type of `tr_globalIPv6()`, that change would touch two other callers and probably make sense in a followup -- but it would remove a lot of hardcoded `16`s :smile:

Reply:
This PR is quite large already, so I'm voting for the next time.